hail/python/hailtop/aiocloud/aiogoogle/client/storage_client.py

```diff
@@ -639,22 +640,76 @@ def schemes() -> Set[str]:
-            "multi_regional",
-        )
+        is_hot_storage = False
+        hot_storage_classes = {"standard", "regional", "multi_regional"}
+        errors: List[aiohttp.ClientResponseError] = []
+        location = self.storage_location(uri)
+
+        async def is_bucket_hot() -> bool:
+            try:
+                return (await self._storage_client.bucket_info(location))["storageClass"].lower() in hot_storage_classes
+            except aiohttp.ClientResponseError as e:
+                errors.append(e)
+                return False
+
+        async def is_object_hot() -> bool:
+            try:
+                return (await (await self.statfile(uri))["storageClass"]).lower() in hot_storage_classes
+            except aiohttp.ClientResponseError as e:
+                errors.append(e)
+                return False
+            except FileNotFoundError:
+                return False
+
+        async def is_dir_first_object_hot() -> bool:
+            try:
+                files = await self.listfiles(uri, recursive=True)
+                next_file = await files.__anext__()
+                while await next_file.is_dir():
+                    next_file = await files.__anext__()
+                return await self.is_hot_storage(await next_file.url())
+            except aiohttp.ClientResponseError as e:
+                errors.append(e)
+                return False
+            except (FileNotFoundError, StopAsyncIteration):
+                return False
+
+        if location not in self.allowed_storage_locations:
+            is_hot_storage = (await is_bucket_hot()) or (await is_object_hot()) or (await is_dir_first_object_hot())
+            if not is_hot_storage:
+                if len(errors) == 0:
+                    raise ValueError(
+                        dedent(f"""\
+                            GCS Bucket '{location}' is configured to use cold storage by default. Accessing the blob
+                            '{uri}' would incur egress charges. Either
+
+                            * avoid the increased cost by changing the default storage policy for the bucket
+                              (https://cloud.google.com/storage/docs/changing-default-storage-class) and the individual
+                              blobs in it (https://cloud.google.com/storage/docs/changing-storage-classes) to 'Standard', or
+
+                            * accept the increased cost by adding '{location}' to the 'gcs_bucket_allow_list' configuration
+                              variable (https://hail.is/docs/0.2/configuration_reference.html).
+                            """)
+                    )
+                error = errors.pop()
+                while len(errors) > 0:
+                    next_error = errors.pop()
+                    next_error.__cause__ = error
+                    error = next_error
+                raise error
+            self.allowed_storage_locations.append(location)
+        return is_hot_storage
```

Comment:
I think this returns `False` when `location in self.allowed_storage_locations`. How come you moved the `allowed_storage_locations` check into this method? I think it was more predictable when that check was in the `validate_file` function and this method was left to just do the storage class check.

Reply:
good catch on the first part, fixed!

for the second, i moved the check into the method because i needed the original `uri`, not the `location`, for `is_object_hot` and `is_dir_first_object_hot`, and didn't want to duplicate the work of getting the `location` from the `uri` across both methods; happy to move it back though if that would still be preferable, it's not a lot of work to duplicate, i mostly was just worried about it affecting the readability of what i'm doing

Reply:
Ya I think that when it comes to readability I care a bit less about the duplicated method call as I do about the interpretation of the signature. I find it surprising that a method called `is_hot_storage` would side-effect and despite being `-> bool` it looks from this implementation that it's more like `True | ValueError` which I find harder to reason about. I think this revision looks great my only request would be to move that `allowed_storage_locations` check along with the big error message up into the `validate_file` function.

Reply:
updated!